tools/2024-06-gitops-feature-spec.md

```diff
@@ -0,0 +1,345 @@
+
+1. Using my editor of choice, VSCode, I'll make changes to the replica number for the pods through changing the parameter of replica sets that my MongoDB recipe takes in as defined in its Radius Environment resource in `env.bicep`:
+
+```bicep
+resource env 'Applications.Core/environments@2023-10-01-preview' = {
+    name: 'prod'
+    properties: {
+      compute: {
+        kind: 'kubernetes'
+        resourceId: 'self'
+        namespace: 'default'
+      }
+      recipes: {
+        'Applications.Datastores/mongoDB':{
+          'mongoDB-bicep': {
+            templateKind: 'bicep'
+            templatePath: 'https://ghcr.io/USERNAME/recipes/myrecipe:1.1.0'
+            // Optionally set parameters for all resources calling this Recipe
+            parameters: {
+              replicaSet: 10
+            }
+          }
+        }
+      }
+    }
+}
+```
+
+2. I can then commit this file to my repository and have Flux apply it to my Radius Environment.
+
+> As an SRE, I'm very happy with this experience because I did not have to do many steps to apply my parameter patch and while I had to learn some Bicep knowledge and Radius knowledge, I was able to ultimately leverage Flux mechanisms which I'm familiar with to apply my patches to the Kubernetes cluster.
+
+_Requirements resulting from this scenario:_
+- Flux and ArgoCD must be able to detect the file change described above and must be able to read the env.bicep file as required to deploy the Radius app correctly 
+- Radius types and Bicep files must be able to be read by Flux/ArgoCD which currently does not have that ability.
+
+## Key investments
+<!-- List the features required to enable this scenario. -->
+
+### Feature 1: GitOps tooling can detect and understand Radius definition files
+<!-- One or two sentence summary -->
+GitOps tools (beginning with Flux and ArgoCD) must be able to detect the file change in the git repo described above and must be able to read the `env.bicep` file as required to deploy the Radius app correctly.
+
+### Feature 2: GitOps tooling can execute Radius commands
+<!-- One or two sentence summary -->
+Once Radius definition file changes are detected and understood, GitOps tools (beginning with Flux and ArgoCD) must be able to execute deployments for the Radius resources defined in the git repo.
+
+### Feature 3: The GitOps controllers must be aware of the patches applied to the Kubernetes cluster
+<!-- One or two sentence summary -->
+The GitOps controllers must be aware of the patches applied to the Kubernetes cluster and must be able to honor those patched configurations to avoid conflicts with the Radius resource definitions that have been checked in. This should be possible by leveraging the existing Kustomize mechanisms in Flux ([Kustomize controller](https://fluxcd.io/flux/components/kustomize/)) and ArgoCD ([Application controller](https://argo-cd.readthedocs.io/en/stable/developer-guide/architecture/components/#application-controller)).
```

Comment:
> The GitOps controllers must be aware of the patches applied to the Kubernetes cluster and must be able to honor those patched configurations to avoid conflicts with the Radius resource definitions that have been checked in

Let's talk about this 😁